silhouette/app/com/mohiva/play/silhouette/impl/providers/state/CsrfStateItemHandler.scala

```diff
@@ -0,0 +1,167 @@
+package com.mohiva.play.silhouette.impl.providers.state
+
+import javax.inject.Inject
+
+import com.mohiva.play.silhouette.api.util.{ ExtractableRequest, IDGenerator }
+import com.mohiva.play.silhouette.impl.providers.SocialStateItem.ItemStructure
+import com.mohiva.play.silhouette.impl.providers.{ PublishableSocialStateItemHandler, SocialStateItem, SocialStateItemHandler }
+import play.api.libs.json.{ Format, JsValue, Json }
+import CsrfStateItemHandler._
+import com.mohiva.play.silhouette.api.crypto.CookieSigner
+import com.mohiva.play.silhouette.impl.exceptions.OAuth2StateException
+import play.api.mvc.{ Cookie, RequestHeader, Result }
+
+import scala.concurrent.duration._
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.util.{ Failure, Success, Try }
+
+case class CsrfState(value: String) extends SocialStateItem
+
+class CsrfStateItemHandler @Inject() (
+  settings: CsrfStateSettings,
+  idGenerator: IDGenerator,
+  cookieSigner: CookieSigner) extends SocialStateItemHandler with PublishableSocialStateItemHandler {
+
+  /**
+   * The item the handler can handle.
+   */
+  override type Item = CsrfState
+
+  /**
+   * Gets the state item the handler can handle.
+   *
+   * @param ec The execution context to handle the asynchronous operations.
+   * @return The state params the handler can handle.
+   */
+  override def item(implicit ec: ExecutionContext): Future[Item] = {
+    idGenerator.generate.map { CsrfState(_) }
+  }
+
+  /**
+   * Indicates if a handler can handle the given [[SocialStateItem]].
+   *
+   * This method should check if the [[serialize]] method of this handler can serialize the given
+   * unserialized state item.
+   *
+   * @param item The item to check for.
+   * @return [[Some]] casted state item if the handler can handle the given state item, [[None]] otherwise.
+   */
+  override def canHandle(item: SocialStateItem): Option[Item] = item match {
+    case i: Item => Some(i)
+    case _       => None
+  }
+
+  /**
+   * Indicates if a handler can handle the given unserialized state item.
+   *
+   * This method should check if the [[unserialize]] method of this handler can unserialize the given
+   * serialized state item.
+   *
+   * @param item The item to check for.
```

Comment:
@akkie The same rule doesn't hold for type params? Inferring from the way you indented in SocialStateProvider.

```scala
  /**
   * Publishes the state to the client.
   *
   * @param item    The item to publish.
   * @param result  The result to send to the client.
   * @param request The current request.
   * @tparam B The type of the request body.
   * @return The result to send to the client.
```
How about the order between type params and params? Which comes first?

```scala
  /**
   * Authenticates the user and returns the auth information and the user state.
   *
   * Returns either a [[StatefulAuthInfo]] if all went OK or a [[play.api.mvc.Result]] that the controller
   * sends to the browser (e.g.: in the case of OAuth where the user needs to be redirected to the service
   * provider).
   *
   * @tparam S       The type of the user state item.
   * @tparam B       The type of the request body.
   * @param format   The JSON format to the transform the user state into JSON.
   * @param request  The request.
   * @param classTag The class tag for the user state item.
   * @return Either a [[play.api.mvc.Result]] or the [[StatefulAuthInfo]] from the provider.
```